README.md
# PHP Cache Duration
![Banner](https://banners.beyondco.de/Cache%20Duration%20Package.png?theme=dark&packageManager=composer+require&packageName=ajimoti%2Fcache-duration&pattern=charlieBrown&style=style_1&description=A+more+readable+way+to+get+time+in+seconds+while+caching&md=1&showWatermark=1&fontSize=100px&images=clock)

## Introduction
A readable and fluent way to generate PHP cache time.

Built and written by [Ajimoti Ibukun](https://www.linkedin.com/in/ibukun-ajimoti-3420a786/)

## Quick Samples
Instead of this:
```php
$cacheDuration = 25 * 60 * 60; // twenty five hours
Redis::expire($userData, $cacheDuration);
```

You can do this
```php
$cacheDuration = Duration::twentyFiveHours(); // returns 25 hours in seconds (25 * 60 * 60)
Redis::expire($userData, $cacheDuration);

// or
$cacheDuration = Duration::hours(25); // returns 25 hours in seconds (25 * 60 * 60)
Redis::expire($userData, $cacheDuration);
```

You can also do this:
```php
$cacheDuration = Duration::at('first day of January 2023'); // returns the time difference between the present time and the first of january 2023 in seconds

Redis::expire($userData, $cacheDuration);
```

## Requirements
- PHP 8.0 or higher

## Installation
You can install the package via composer:
```bash
composer require ajimoti/cache-duration --with-all-dependencies
```

## Documentation
After installing the package via composer, import the `Duration` trait inside your class, then you are set.
```php
<?php
require 'vendor/autoload.php';

use Ajimoti\CacheDuration\Duration;

var_dump(Duration::fourtyMinutes()); // returns 2400;
var_dump(Duration::tenHours()); // returns 36000;
var_dump(Duration::fiftyFourDays()); // returns 4665600;
```

### Available methods
| Method      | Expectations | 
| ----------- | ----------- |
| [seconds($value)](#secondsvalue)  | Expects time in seconds  |
| [minutes($value)](#minutesvalue)   | Expects time in minutes  |
| [hours($value)](#hoursvalue)  | Expects time in hours  |
| [days($value)](#daysvalue) | Expects time in days  |
| [at($value)](#atvalue)  | Expects `string`, `carbon` instance, or `DateTime` instance  |

### Dynamic calls
In addition to the methods provided above, the package uses `PHP` `__callStatic()` method to allow you make dynamic calls on the `Duration` trait.

For example, you want to get the number of seconds in 37 days, you can achieve this by calling a `camel-case` text of the number (`thirtySeven` in this case), plus the unit (`Days` in this case). That will leave us with something like this:

```php
// The formula = camelCaseOfTheNumberInWords + Unit
Duration::thirtySevenDays(); // returns the number of seconds in 37 days
```

> **Note:** The number in words **MUST** be in `camel-case`. Any other case will throw an `InvalidArgumentException`. Additionally, it must be followed by a `title-case` of the unit. The available units are `Seconds`, `Minutes`, `Hours`, and `Days`.


## Usage
### `seconds($value)`
Get time in seconds. It basically returns the same value passed into it.
```php
use Ajimoti\CacheDuration\Duration;

$cacheDuration = Duration::seconds(30); // returns 30

// or dynamically
$cacheDuration = Duration::thirtySeconds(); // returns 30
```

### `minutes($value)`
Converts time in minutes into seconds.
```php
use Ajimoti\CacheDuration\Duration;

$cacheDuration = Duration::minutes(55); // returns 55 minutes in seconds (55 * 60)

// or dynamically
$cacheDuration = Duration::fiftyFiveMinutes(); // returns 55 minutes in seconds (55 * 60)
```

### `hours($value)`
Converts time in hours into seconds.
```php
use Ajimoti\CacheDuration\Duration;

$cacheDuration = Duration::hours(7); // returns 7 hours in seconds (7 * 60 * 60)

// or dynamically
$cacheDuration = Duration::sevenHours(); // returns 7 hours in seconds (7 * 60 * 60)
```

### `days($value)`
Converts time in days into seconds.
```php
use Ajimoti\CacheDuration\Duration;

$cacheDuration = Duration::days(22); // returns 22 days in seconds (22 * 24 * 60 * 60)

// or dynamically
$cacheDuration = Duration::twentyTwoDays(); // returns 22 days in seconds (22 * 24 * 60 * 60)
```

### `at($value)`
This method allows you to convert a `Carbon\Carbon` instance, `DateTime` instance or  `string` of date into seconds. 

It returns the difference in seconds between the argument passed and the current `timestamp`.

> The date passed into this method **MUST** be a date in the future. When a string is passed, the text **MUST** be compatible with `Carbon::parse()` method, else an exception will be thrown

#### Examples
```php
use Date;
use Carbon\Carbon;
use Ajimoti\CacheDuration\Duration;

// Carbon instance
$cacheDuration = Duration::at(Carbon::now()->addMonths(3)); // returns time in seconds between the present timestamp and three months time

// Datetime instance
$cacheDuration = Duration::at(new DateTime('2039-09-30')); // returns time in seconds between the present timestamp and the date passed (2039-09-30).

// String
$cacheDuration = Duration::at('first day of January 2023'); // returns time in seconds between the present timestamp and the first of January 2023.
```

## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](.github/CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

- [ajimoti](https://github.com/ajimoti)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `php-timer` will be documented in this file.

## 1.0.0 - 202X-XX-XX

- initial release


tests/TimeTest.php
<?php

use Ajimoti\CacheTime\Time;
use Carbon\Carbon;

it('can return time in seconds', function () {
    expect(Time::seconds(1))->toBe(1);
});

it('can convert minutes to seconds', function () {
    $randomMinutes = random_int(1, 100);
    expect(Time::minutes($randomMinutes))->toBe(60 * $randomMinutes);
});

it('can convert hours to seconds', function () {
    $randomHours = random_int(1, 100);
    expect(Time::hours($randomHours))->toBe(60 * 60 * $randomHours);
});

it('can convert days to seconds', function () {
    $randomDay = random_int(1, 100);
    expect(Time::days($randomDay))->toBe(60 * 60 * 24 * $randomDay);
});

it('can calculate the difference in seconds between now and a provided date', function () {
    $twoWeeksTime = Carbon::now()->addWeeks(2);

    expect(Time::at($twoWeeksTime))->toBe(Carbon::now()->diffInSeconds($twoWeeksTime));
});

it('at method can accept a string', function () {
    $year = (int) date('Y') + 1;

    $sentence = "first day of January {$year}";

    expect(Time::at("{$year}-01-01"))->toBe(Carbon::createMidnightDate($year, 01, 01)->diffInSeconds(Carbon::now()));
    expect(Time::at($sentence))->toBe(Carbon::parse($sentence)->diffInSeconds(Carbon::now()));
});

it('at method accepts a string with seconds', function () {
    $year = (int) date('Y') + 1;
    $sentence = "first day of January {$year} 00:00:00";

    expect(Time::at($sentence))->toBe(Carbon::parse($sentence)->diffInSeconds(Carbon::now()));
});

it('at method accepts dateTime instance', function () {
    $year = (int) date('Y') + 1;
    $sentence = "first day of January {$year}";

    expect(Time::at(new \DateTime($sentence)))->toBe(Carbon::parse($sentence)->diffInSeconds(Carbon::now()));
});

it('at method accepts carbon instance', function () {
    $carbon = Carbon::now()->addMonths(2);

    expect(Time::at($carbon))->toBe($carbon->diffInSeconds(Carbon::now()));
});

it('at method throws an exception when a wrong type is passed', function () {
    Time::at(1);
})->throws(InvalidArgumentException::class);


tests/DurationTest.php
<?php

use Ajimoti\CacheDuration\Duration;
use Carbon\Carbon;

it('can return time in seconds', function () {
    expect(Duration::seconds(1))->toBe(1);
});

it('can convert minutes to seconds', function () {
    $randomMinutes = random_int(1, 100);
    expect(Duration::minutes($randomMinutes))->toBe(60 * $randomMinutes);
});

it('can convert hours to seconds', function () {
    $randomHours = random_int(1, 100);
    expect(Duration::hours($randomHours))->toBe(60 * 60 * $randomHours);
});

it('can convert days to seconds', function () {
    $randomDay = random_int(1, 100);
    expect(Duration::days($randomDay))->toBe(60 * 60 * 24 * $randomDay);
});

it('can calculate the difference in seconds between now and a provided date', function () {
    $twoWeeksTime = Carbon::now()->addWeeks(2);

    expect(Duration::at($twoWeeksTime))->toBe(Carbon::now()->diffInSeconds($twoWeeksTime));
});

it('at method can accept a string', function () {
    $year = (int) date('Y') + 1;

    $sentence = "first day of January {$year}";

    expect(Duration::at("{$year}-01-01"))->toBe(Carbon::createMidnightDate($year, 01, 01)->diffInSeconds(Carbon::now()));
    expect(Duration::at($sentence))->toBe(Carbon::parse($sentence)->diffInSeconds(Carbon::now()));
});

it('at method accepts a string with seconds', function () {
    $year = (int) date('Y') + 1;
    $sentence = "first day of January {$year} 00:00:00";

    expect(Duration::at($sentence))->toBe(Carbon::parse($sentence)->diffInSeconds(Carbon::now()));
});

it('at method accepts dateTime instance', function () {
    $year = (int) date('Y') + 1;
    $sentence = "first day of January {$year}";

    expect(Duration::at(new \DateTime($sentence)))->toBe(Carbon::parse($sentence)->diffInSeconds(Carbon::now()));
});

it('at method accepts carbon instance', function () {
    $carbon = Carbon::now()->addMonths(2);

    expect(Duration::at($carbon))->toBe($carbon->diffInSeconds(Carbon::now()));
});

it('at method throws an exception when a wrong type is passed', function () {
    Duration::at(1);
})->throws(InvalidArgumentException::class);


tests/Units/StrTest.php
<?php

use Ajimoti\CacheDuration\Helpers\Str;

it('can correctly pluralize method names', function () {
    $singleToPlural = [
        'second' => 'seconds',
        'minute' => 'minutes',
        'hour' => 'hours',
        'day' => 'days',
    ];

    foreach ($singleToPlural as $single => $plural) {
        expect(Str::simplePluralize($single))->toBe($plural);
    }
});

it('can correctly convert words to numbers', function () {
    $wordsToNumbers = [
        'one' => 1,
        'two' => 2,
        'three' => 3,
        'four' => 4,
        'twenty' => 20,
        'twenty three' => 23,
        'thirty' => 30,
        'thirty six' => 36,
        'forty' => 40,
        'forty four' => 44,
        'fifty' => 50,
        'fifty one' => 51,
        'fifty two' => 52,
        'sixty three' => 63,
    ];

    foreach ($wordsToNumbers as $word => $number) {
        expect(Str::wordsToNumber($word))->toBe($number);
    }
});

it('can correctly convert camel to space separated', function () {
    $camelCaseToSpaceSeparated = [
        'thisIsALongWord' => 'this is a long word',
        'thisIsALongWordWithNumbers' => 'this is a long word with numbers',
        'thisIsALongWordWithNumbersAndSymbols' => 'this is a long word with numbers and symbols',
        'iNeedAJob' => 'i need a job',
    ];

    foreach ($camelCaseToSpaceSeparated as $camel => $space) {
        expect(Str::camelToSpaceSeparated($camel))->toBe($space);
    }
});


src/Time.php
<?php

namespace Ajimoti\CacheTime;

use Ajimoti\CacheTime\Traits\SupportsDynamicCalls;
use Carbon\Carbon;
use DateTime;
use InvalidArgumentException;

final class Time
{
    use SupportsDynamicCalls;

    /**
     * Returns the time in seconds.
     *
     * @param int $seconds
     */
    final public static function seconds(int $seconds = 1): int
    {
        return $seconds;
    }

    /**
     * Returns the number of minutes in seconds.
     *
     * @param int $minutes
     */
    final public static function minutes(int $minutes = 1): int
    {
        return static::seconds(60) * $minutes;
    }

    /**
     * Returns the number of hours in seconds.
     *
     * @param int $hours
     */
    final public static function hours(int $hours = 1): int
    {
        return static::minutes(60) * $hours;
    }

    /**
     * Returns the number of days in seconds.
     *
     * @param int $hours
     */
    final public static function days(int $days = 1): int
    {
        return static::hours(24) * $days;
    }

    /**
     * Returns the number of seconds between now and the time provided.
     *
     * @param DateTime|Carbon|string $value - string can be HH:MM:SS or HH:MM
     */
    final public static function at(DateTime|Carbon|string $value): int
    {
        if ($value instanceof Carbon) {
            $now = Carbon::now();

            if ($value->lte($now)) {
                throw new InvalidArgumentException("The value provided [{$value}] must be greater than [{$now}].");
            }

            return $value->diffInSeconds(Carbon::now());
        }

        if ($value instanceof DateTime) {
            return static::at(Carbon::instance($value));
        }

        if (is_string($value)) {
            return static::at(Carbon::parse($value));
        }

        throw new InvalidArgumentException('Invalid value provided.');
    }

    /**
     * Handle dynamic static method calls on the class.
     *
     * @var int
     */
    public static function __callStatic($methodName, $arguments)
    {
        return static::dynamicCall($methodName);
    }
}


src/Duration.php
<?php

namespace Ajimoti\CacheDuration;

use Ajimoti\CacheDuration\Traits\SupportsDynamicCalls;
use Carbon\Carbon;
use DateTime;
use InvalidArgumentException;

final class Duration
{
    use SupportsDynamicCalls;

    /**
     * Returns the time in seconds.
     *
     * @param int $seconds
     */
    final public static function seconds(int $seconds = 1): int
    {
        return $seconds;
    }

    /**
     * Returns the number of minutes in seconds.
     *
     * @param int $minutes
     */
    final public static function minutes(int $minutes = 1): int
    {
        return static::seconds(60) * $minutes;
    }

    /**
     * Returns the number of hours in seconds.
     *
     * @param int $hours
     */
    final public static function hours(int $hours = 1): int
    {
        return static::minutes(60) * $hours;
    }

    /**
     * Returns the number of days in seconds.
     *
     * @param int $hours
     */
    final public static function days(int $days = 1): int
    {
        return static::hours(24) * $days;
    }

    /**
     * Returns the number of seconds between now and the time provided.
     *
     * @param DateTime|Carbon|string $value - string can be HH:MM:SS or HH:MM
     */
    final public static function at(DateTime|Carbon|string $value): int
    {
        if ($value instanceof Carbon) {
            $now = Carbon::now();

            if ($value->lte($now)) {
                throw new InvalidArgumentException("The value provided [{$value}] must be greater than [{$now}].");
            }

            return $value->diffInSeconds(Carbon::now());
        }

        if ($value instanceof DateTime) {
            return static::at(Carbon::instance($value));
        }

        if (is_string($value)) {
            return static::at(Carbon::parse($value));
        }

        throw new InvalidArgumentException('Invalid value provided.');
    }

    /**
     * Handle dynamic static method calls on the class.
     *
     * @var int
     */
    public static function __callStatic($methodName, $arguments)
    {
        return static::dynamicCall($methodName);
    }
}


src/Traits/SupportsMagicCalls.php
<?php

namespace Ajimoti\Timer\Traits;

use Ajimoti\Timer\Exceptions\BadMethodException;
use Ajimoti\Timer\Exceptions\InvalidMethodNameException;
use Ajimoti\Timer\Helpers\Str;

trait SupportsMagicCalls
{
    private static function allowedSuffixes(): array
    {
        $allowedSuffixes = ['second', 'minute', 'hour', 'day'];

        foreach ($allowedSuffixes as $allowedSuffix) {
            $allowedSuffixes[] = Str::simplePluralize($allowedSuffix);
        }

        return $allowedSuffixes;
    }

    public static function doMagic($methodName)
    {
        $method = null;
        foreach (static::allowedSuffixes() as $suffix) {
            if (Str::endsWith(strtolower($methodName), $suffix)) {
                $value = Str::wordsToNumber(
                    Str::studlyToSpaceSeparated(Str::eraseSuffix($methodName, $suffix))
                );

                $method = match ($suffix) {
                    'second' => 'seconds',
                    'minute' => 'minutes',
                    'hour' => 'hours',
                    'day' => 'days',
                    default => $suffix
                };

                break;
            }
        }

        if (empty($method)) {
            throw new BadMethodException(static::class, $methodName);
        }

        if (isset($method) && empty($value)) {
            throw new InvalidMethodNameException($methodName);
        }

        return static::$method($value);
    }
}


src/Traits/SupportsDynamicCalls.php
<?php

namespace Ajimoti\CacheDuration\Traits;

use Ajimoti\CacheDuration\Exceptions\BadMethodException;
use Ajimoti\CacheDuration\Helpers\Str;

trait SupportsDynamicCalls
{
    /**
     * Get magic calls allowed suffixes.
     *
     * @var array
     */
    private static function allowedSuffixes(): array
    {
        $allowedSuffixes = ['second', 'minute', 'hour', 'day'];

        foreach ($allowedSuffixes as $allowedSuffix) {
            $allowedSuffixes[] = Str::simplePluralize($allowedSuffix);
        }

        return $allowedSuffixes;
    }

    /**
     * Handle dynamic method calls on the class.
     *
     * @var int
     */
    public static function dynamicCall($methodName): int
    {
        $wordsInMethodName = explode(' ', Str::camelToSpaceSeparated($methodName));
        $suffix = end($wordsInMethodName);

        if (! in_array($suffix, static::allowedSuffixes())) {
            throw new BadMethodException(static::class, $methodName);
        }

        $value = Str::wordsToNumber(
            implode(' ', array_diff($wordsInMethodName, [$suffix]))
        );

        $method = match ($suffix) {
            'second' => 'seconds',
            'minute' => 'minutes',
            'hour' => 'hours',
            'day' => 'days',
            default => $suffix
        };

        return static::$method($value);
    }
}


src/Helpers/Str.php
<?php

namespace Ajimoti\CacheDuration\Helpers;

use InvalidArgumentException;
use SplStack;

class Str
{
    /**
     * The cache of words that will be converted to number.
     *
     * @var array
     */
    protected static $wordsToNumberCache = [];

    /**
     * Get the plural of a string.
     *
     * Basically appends an `s` to argument provided
     *
     * @param string $word
     *
     * @return string
     */
    public static function simplePluralize(string $word): string
    {
        return $word . 's';
    }

    /**
    * Convert a string such as "one hundred thousand" to 100000
    * This is useful for converting strings like "one hundred thousand" to numbers.
    *
    * Credit: https://stackoverflow.com/a/11219737/16163383
    *
    * @param string $data The numeric string.
    *
    * @return int The numeric value.
    */
    public static function wordsToNumber($data): int
    {
        $key = $data;
        if (isset(static::$wordsToNumberCache[$key])) {
            return static::$wordsToNumberCache[$key];
        }

        $wordsToValue = [
            'zero' => '0',
            'a' => '1',
            'one' => '1',
            'two' => '2',
            'three' => '3',
            'four' => '4',
            'five' => '5',
            'six' => '6',
            'seven' => '7',
            'eight' => '8',
            'nine' => '9',
            'ten' => '10',
            'eleven' => '11',
            'twelve' => '12',
            'thirteen' => '13',
            'fourteen' => '14',
            'fifteen' => '15',
            'sixteen' => '16',
            'seventeen' => '17',
            'eighteen' => '18',
            'nineteen' => '19',
            'twenty' => '20',
            'thirty' => '30',
            'forty' => '40',
            'fourty' => '40', // common misspelling
            'fifty' => '50',
            'sixty' => '60',
            'seventy' => '70',
            'eighty' => '80',
            'ninety' => '90',
            'hundred' => '100',
            'thousand' => '1000',
            'million' => '1000000',
            'billion' => '1000000000',
            'and' => '',
        ];

        $data = strtolower($data);
        $validWords = array_keys($wordsToValue);

        // Ensure every word provided is valid
        foreach (explode(' ', $data) as $word) {
            if (! in_array($word, $validWords)) {
                throw new InvalidArgumentException(sprintf(
                    'Invalid word "%s" in number "%s". Ensure words are in camel case format',
                    $word,
                    $data
                ));
            }
        }

        // Replace all number words with an equivalent numeric value
        $data = strtr($data, $wordsToValue);

        // Coerce all tokens to numbers
        $parts = array_map(
            function ($val) {
                return floatval($val);
            },
            preg_split('/[\s-]+/', $data)
        );

        $stack = new SplStack(); // Current work stack
        $sum = 0; // Running total
        $last = null;

        foreach ($parts as $part) {
            if (! $stack->isEmpty()) {
                // We're part way through a phrase
                if ($stack->top() > $part) {
                    // Decreasing step, e.g. from hundreds to ones
                    if ($last >= 1000) {
                        // If we drop from more than 1000 then we've finished the phrase
                        $sum += $stack->pop();
                        // This is the first element of a new phrase
                        $stack->push($part);
                    } else {
                        // Drop down from less than 1000, just addition
                        // e.g. "seventy one" -> "70 1" -> "70 + 1"
                        $stack->push($stack->pop() + $part);
                    }
                } else {
                    // Increasing step, e.g ones to hundreds
                    $stack->push($stack->pop() * $part);
                }
            } else {
                // This is the first element of a new phrase
                $stack->push($part);
            }

            // Store the last processed part
            $last = $part;
        }

        return static::$wordsToNumberCache[$key] = (int) ($sum + $stack->pop());
    }

    /**
     * Convert a camel case string to space separated words.
     *
     * @param string $string
     *
     * @return string
     */
    public static function camelToSpaceSeparated(string $string): string
    {
        $words = array_map(function ($word) {
            return strtolower($word);
        }, preg_split('/(?=[A-Z])/', $string));

        return implode(' ', $words);
    }
}
